src/textual/document/_document.py

```diff
@@ -0,0 +1,366 @@
+
+        Args:
+            start: The start location of the selection.
+            end: The end location of the selection.
+
+        Returns:
+            The text between start (inclusive) and end (exclusive).
+        """
+
+    @abstractmethod
+    def get_size(self, indent_width: int) -> Size:
+        """Get the size of the document.
+
+        The height is generally be the number of lines, and the width
+        is generally the maximum cell length of all the lines.
+
+        Args:
+            indent_width: The width to use for tab characters.
+
+        Returns:
+            The Size of the document bounding box.
+        """
+
+    @property
+    @abstractmethod
+    def line_count(self) -> int:
+        """Returns the number of lines in the document."""
+
+    @overload
+    def __getitem__(self, line_index: int) -> str:
+        ...
+
+    @overload
+    def __getitem__(self, line_index: slice) -> list[str]:
+        ...
+
+    @abstractmethod
+    def __getitem__(self, line_index: int | slice) -> str | list[str]:
+        """Return the content of a line as a string, excluding newline characters.
+
+        Args:
+            line_index: The index or slice of the line(s) to retrieve.
+
+        Returns:
+            The line or list of lines requested.
+        """
+
+
+class Document(DocumentBase):
+    """A document which can be opened in a TextArea."""
+
+    def __init__(self, text: str) -> None:
+        self._newline = _detect_newline_style(text)
+        """The type of newline used in the text."""
+        self._lines: list[str] = text.splitlines(keepends=False)
+        """The lines of the document, excluding newline characters.
+
+        If there's a newline at the end of the file, the final line is an empty string.
+        """
+        if text.endswith(tuple(VALID_NEWLINES)) or not text:
+            self._lines.append("")
+
+    @property
+    def lines(self) -> list[str]:
+        """Get the document as a list of strings, where each string represents a line.
+
+        Newline characters are not included in at the end of the strings.
+
+        The newline character used in this document can be found via the `Document.newline` property.
+        """
+        return self._lines
+
+    @property
+    def text(self) -> str:
+        """Get the text from the document."""
+        return self._newline.join(self._lines)
+
+    @property
+    def newline(self) -> Newline:
+        """Get the Newline used in this document (e.g. '\r\n', '\n'. etc.)"""
+        return self._newline
+
+    def get_size(self, tab_width: int) -> Size:
+        """The Size of the document, taking into account the tab rendering width.
+
+        Args:
+            tab_width: The width to use for tab indents.
+
+        Returns:
+            The size (width, height) of the document.
+        """
+        lines = self._lines
+        cell_lengths = [cell_len(line.expandtabs(tab_width)) for line in lines]
```

Comment:
> If lines is a list of str, then I'm afraid expandtabs won't work for us, as it is unaware of double cell characters. The Rich Text class can expand tabs in a double-cell aware manner.

Good catch - I wonder if it's worth wrapping line in a `Text` instance here or if we should extract `expandtabs` from `Text` such that I can run it against a string without constructing a new `Text` instance.

> This function could be quite inefficient. Might be significant with a large enough document. Is this called a lot, is there room for caching?

This method is top of my list for optimisation. I suspect it's where a majority of the time is spent on pressing a key at the moment, although I haven't confirmed.

Reply:
> Good catch - I wonder if it's worth wrapping line in a `Text` instance here or if we should extract `expandtabs` from `Text` such that I can run it against a string without constructing a new `Text` instance.

We could, yeah. `Text.expandtabs` has to do additional work to handle spans, which we wouldn't need here.

Reply:
Some information from profiling this:

- When no syntax highlighting is used this method `Document.get_size` accounts for about 70% of the time taken when you press a key.
- When syntax highlighting is used, I can't find this method on the profile - I think it's because it's dwarfed by the time taken doing highlighting (93% of the time taken on pressing a key is related to syntax highlighting). `SyntaxAwareDocument._prepare_highlights` should probably be the first thing we target with optimisation.

tl;dr: `Document.get_size` and `SyntaxAwareDocument._prepare_highlights` are the two largest bottlenecks on keypresses.